Add alt PUL&CUL barcodes to Development & Production config

nonword opened the pull request:
This:

 - Proposes merging the barcode reassignment work already merged into master. This work includes:
   - When processing checkout requests, checks incoming `patronBarcode` to determine if it's for the CUL or PUL patron barcode that ReCAP is configured to use
   - If the checkout request is determined to be for the CUL or PUL patrons, new functionality *overrides* the `patronBarcode` with a random barcode from a configurable list of CUL/PUL barcodes
   - Proceeds as before (i.e. forwards request to Sierra)
 - Adds encrypted alt barcode config lines to var_development, var_qa, and var_production. Three alt barcodes have been created in Sierra Test for each inst. 10 have been created in Sierra Prod. In both cases the config val has been encrypted because we couldn't decide if it was sensitive. (Note that encryption seems to be presumed disabled for tests.)

ktp242 commented on src/Controller/CheckoutRequestController.php:
```diff
@@ -252,4 +254,24 @@ protected function processException($errorType, $errorMessage, \Exception $excep
+    public function reassignPartnerBarcode($data)
+    {
+      $key = "PATRON_BARCODES_{$data['patronBarcode']}";
+      $barcodes = explode("," , Config::get($key, "", true));
```
I think a better empty value for the second parameter probably should be `null`. It fits the definition of the function `get` in `NYPL\Starter\Config`

```
public static function get($name = '', $defaultValue = null, $isEncrypted = false)
```

nonword replied:
I def prefer `null` over an empty value, but in this case we're going to call `explode` on that default value. PHP's `explode` seems to interpret splitting `null` as being equivalent to splitting `''`, but that seems less clear to me (i.e. it surprises me that splitting `null` behaves this way). I think given that the [official doc on `explode`](http://php.net/explode) declares the param a `string`, we should probably send it a string so that we know what we're going to get back.

ktp242 commented on src/Controller/CheckoutRequestController.php:
```diff
@@ -252,4 +254,24 @@ protected function processException($errorType, $errorMessage, \Exception $excep
+      $numberOfPatronBarcodes = count($barcodes);
+      if ($numberOfPatronBarcodes >= 1) {
+        $newBarcode = $barcodes[rand(0, $numberOfPatronBarcodes - 1)];
+        if ($newBarcode) {
```
I like the check here to see if the data really exits! But I think a better place will be line #265. Because if for any reason `Config::get` failed, it will return `null` by default, or since we put `""` here, it will now return `""`.  And then it will throw warning: ` Undefined offset: 1 in...` from `explode ()`

nonword replied:
If `Config::get` falls back on the default, we'll run `explode` on an empty string (or `null` if we change that to `null`). In either case, `$barcodes` will be an array of 0 - N length. Only if length is greater than 0 will we attempt to reach into the array by offset. Having trouble seeing where the error might be thrown.. Where might an `Undefined offset` error be thrown from?

ktp242 commented:
I tried to run `npm run test` and realized that I need to install `phplint` first, and then it threw an error about missing `config/var_testing.env`. Probably we can add some documentations for running the tests?

nonword commented:
@ktp242 thanks for testing so carefully. We overlooked things. I've pushed some commits that hopefully clarify testing. The official test suite is in PHP, so not invoked via `npm test`. I've updated documentation and wired `package.json` to run the official php unit tests. Also I added `phplint` as a dev dependency.
